File: frontend/src/api/pipl.ts
/**
 * PIPL 数据主体权利 API（对齐 D05 §13，/api/v1/pipl/* 命名空间）
 * 对齐 baseline §7.1
 */
import { get, post } from './request'
import type { ConsentStatus, ConsentPurpose } from '@/types/enum'

/** 同意记录（D05 §13.1） */
export interface ConsentRecord {
  consent_id: string
  user_id: string
  purpose: ConsentPurpose
  legal_basis: string
  consent_type: 'EXPLICIT' | 'IMPLICIT_BY_ACTION'
  status: ConsentStatus
  scope: string[]
  granted_at: string
  expires_at?: string
  withdrawn_at?: string
  policy_version: string
  evidence_ref?: string
}

/** 同意列表响应（D05 §13.3） */
export interface ConsentListResult {
  user_id: string
  items: ConsentRecord[]
  summary: { active_count: number; withdrawn_count: number; expired_count: number }
  page: number
  page_size: number
  total: number
}

/** 数据导出任务（D05 §13.4-13.5） */
export interface DataExportTask {
  task_id: string
  user_id: string
  status: 'PROCESSING' | 'READY' | 'FAILED' | 'EXPIRED'
  scope?: string[]
  format?: 'JSON' | 'CSV' | 'XLSX'
  download_url?: string
  expires_at?: string
  estimated_seconds?: number
  created_at?: string
  completed_at?: string
  callback_event?: string
}

/** 删除请求（D05 §13.6-13.7） */
export interface DeletionRequest {
  request_id: string
  user_id: string
  status: 'PENDING_REVIEW' | 'APPROVED' | 'IN_PROGRESS' | 'COMPLETED' | 'REJECTED' | 'PARTIALLY_COMPLETED'
  scope: string[]
  deleted_count?: number
  anonymized_count?: number
  retained_count?: number
  retention_reason?: string
  legal_hold_review?: { required: boolean; reviewer_id?: string; reviewed_at?: string }
  estimated_seconds?: number
  created_at?: string
  completed_at?: string
  callback_event?: string
}

/** 更正请求（D05 §13.8） */
export interface RectificationRequest {
  request_id: string
  user_id: string
  status: 'PENDING_REVIEW' | 'APPROVED' | 'IN_PROGRESS' | 'COMPLETED' | 'REJECTED' | 'PARTIALLY_COMPLETED'
  correction_count: number
  estimated_seconds?: number
  callback_event?: string
}

/** 记录同意（D05 §13.1） */
export function grantConsent(payload: {
  user_id: string
  verification_token: string
  consent_type: 'EXPLICIT' | 'IMPLICIT_BY_ACTION'
  purpose: ConsentPurpose
  legal_basis: string
  scope: string[]
  policy_version: string
  expires_at?: string
  evidence: { channel: string; user_agent: string; ip_address: string; signed_text_hash: string }
}) {
  return post<ConsentRecord>('/pipl/consent', payload, {
    headers: { 'X-Idempotency-Key': crypto.randomUUID() }
  })
}

/** 撤回同意（D05 §13.2） */
export function withdrawConsent(payload: {
  user_id: string
  verification_token: string
  consent_id: string
  withdrawal_reason?: 'NO_LONGER_NEEDED' | 'SERVICE_CANCELLED' | 'PRIVACY_CONCERN' | 'OTHER'
  effective_immediately?: boolean
}) {
  return post<ConsentRecord>('/pipl/consent/withdraw', payload)
}

/** 查询用户同意状态（D05 §13.3） */
export function getConsent(
  userId: string,
  params: { purpose?: ConsentPurpose; status?: ConsentStatus; include_history?: boolean } = {}
) {
  return get<ConsentListResult>(`/pipl/consent/${userId}`, params as Record<string, unknown>)
}

/** 申请数据导出（D05 §13.4） */
export function requestDataExport(params: {
  user_id: string
  verification_token: string
  scope: string
  format?: 'JSON' | 'CSV' | 'XLSX'
  start_date?: string
  end_date?: string
  delivery_method?: 'OSS_PRESIGNED_URL' | 'WEBHOOK'
}) {
  return get<DataExportTask>('/pipl/data-export', params as Record<string, unknown>)
}

/** 查询导出状态（D05 §13.5） */
export function getDataExportStatus(taskId: string) {
  return get<DataExportTask>(`/pipl/data-export/${taskId}/status`)
}

/** 申请数据删除（D05 §13.6） */
export function requestDeletion(payload: {
  user_id: string
  verification_token: string
  scope: string[]
  reason: 'USER_REQUEST' | 'CONSENT_WITHDRAWN' | 'DATA_RETENTION_EXPIRED' | 'LEGAL_OBLIGATION_END'
  retain_for_aml?: boolean
  legal_hold_review?: boolean
}) {
  return post<DeletionRequest>('/pipl/deletion', payload, {
    headers: { 'X-Idempotency-Key': crypto.randomUUID() }
  })
}

/** 查询删除状态（D05 §13.7） */
export function getDeletionStatus(requestId: string) {
  return get<DeletionRequest>(`/pipl/deletion/${requestId}/status`)
}

/** 数据更正请求（D05 §13.8） */
export function requestRectification(payload: {
  user_id: string
  verification_token: string
  corrections: {
    resource_type: 'TRANSACTION' | 'USER_PROFILE' | 'CASE' | 'CONSENT'
    resource_id: string
    field: string
    current_value: unknown
    corrected_value: unknown
    evidence: string
  }[]
  reason: 'DATA_INACCURATE' | 'DATA_OUTDATED' | 'DATA_INCOMPLETE' | 'USER_REQUEST'
}) {
  return post<RectificationRequest>('/pipl/rectification', payload, {
    headers: { 'X-Idempotency-Key': crypto.randomUUID() }
  })
}

File: backend/app/schemas/pipl.py
"""PIPL schemas（D05 §13）。"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class ConsentTypeEnum(str, Enum):
    EXPLICIT = "EXPLICIT"
    IMPLICIT_BY_ACTION = "IMPLICIT_BY_ACTION"


class ConsentPurpose(str, Enum):
    """同意用途（基准 §3.11）。"""

    TRANSACTION_SCORING = "TRANSACTION_SCORING"
    FRAUD_DETECTION = "FRAUD_DETECTION"
    AML_REPORT = "AML_REPORT"
    MARKETING = "MARKETING"
    RESEARCH = "RESEARCH"


class ConsentStatus(str, Enum):
    """同意状态（基准 §3.11）。"""

    GRANTED = "GRANTED"
    WITHDRAWN = "WITHDRAWN"
    EXPIRED = "EXPIRED"


class LegalBasis(str, Enum):
    CONSENT = "CONSENT"
    CONTRACT = "CONTRACT"
    LEGAL_OBLIGATION = "LEGAL_OBLIGATION"
    VITAL_INTEREST = "VITAL_INTEREST"
    PUBLIC_TASK = "PUBLIC_TASK"
    LEGITIMATE_INTEREST = "LEGITIMATE_INTEREST"


class ConsentCreate(BaseModel):
    """POST /pipl/consent 请求体（D05 §13.1）。"""

    user_id: str
    verification_token: str
    consent_type: ConsentTypeEnum
    purpose: ConsentPurpose
    legal_basis: LegalBasis
    scope: list[str]
    policy_version: str
    expires_at: datetime | None = None
    evidence: dict[str, Any]


class ConsentOut(BaseModel):
    """同意记录详情。"""

    consent_id: str
    user_id: str
    status: ConsentStatus
    purpose: ConsentPurpose
    legal_basis: LegalBasis
    consent_type: ConsentTypeEnum
    scope: list[str] = Field(default_factory=list)
    granted_at: datetime | None = None
    expires_at: datetime | None = None
    withdrawn_at: datetime | None = None
    policy_version: str
    evidence_ref: str | None = None


class ConsentWithdraw(BaseModel):
    """POST /pipl/consent/withdraw 请求体（D05 §13.2）。"""

    user_id: str
    verification_token: str
    consent_id: str
    withdrawal_reason: str | None = None
    effective_immediately: bool = True


class DataExportRequest(BaseModel):
    """GET /pipl/data-export 请求参数（D05 §13.4）。"""

    user_id: str
    verification_token: str
    scope: str
    format: str = "JSON"
    start_date: str | None = None
    end_date: str | None = None
    delivery_method: str = "OSS_PRESIGNED_URL"


class DataExportStatusOut(BaseModel):
    """GET /pipl/data-export/{task_id}/status 响应（D05 §13.5）。"""

    task_id: str
    user_id: str
    status: str
    scope: list[str] = Field(default_factory=list)
    format: str = "JSON"
    download_url: str | None = None
    expires_at: datetime | None = None
    created_at: datetime | None = None
    completed_at: datetime | None = None


class DeletionRequestIn(BaseModel):
    """POST /pipl/deletion 请求体（D05 §13.6）。"""

    user_id: str
    verification_token: str
    scope: list[str]
    reason: str
    retain_for_aml: bool = True
    legal_hold_review: bool = True


class DeletionStatusOut(BaseModel):
    """GET /pipl/deletion/{request_id}/status 响应（D05 §13.7）。"""

    request_id: str
    user_id: str
    status: str
    scope: list[str] = Field(default_factory=list)
    deleted_count: int = 0
    anonymized_count: int = 0
    retained_count: int = 0
    retention_reason: str | None = None
    created_at: datetime | None = None
    completed_at: datetime | None = None


class RectificationItem(BaseModel):
    """更正项。"""

    resource_type: str
    resource_id: str
    field: str
    current_value: Any
    corrected_value: Any
    evidence: str


class RectificationRequest(BaseModel):
    """POST /pipl/rectification 请求体（D05 §13.8）。"""

    user_id: str
    verification_token: str
    corrections: list[RectificationItem]
    reason: str


class RectificationStatusOut(BaseModel):
    """更正请求状态。"""

    request_id: str
    user_id: str
    status: str
    correction_count: int
    estimated_seconds: int = 120


__all__ = [
    "ConsentCreate",
    "ConsentOut",
    "ConsentStatus",
    "ConsentTypeEnum",
    "ConsentPurpose",
    "ConsentWithdraw",
    "DataExportRequest",
    "DataExportStatusOut",
    "DeletionRequestIn",
    "DeletionStatusOut",
    "LegalBasis",
    "RectificationItem",
    "RectificationRequest",
    "RectificationStatusOut",
]
